service/resource/dnsrecord/delete.go

```diff
@@ -0,0 +1,85 @@
+package dnsrecord
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/arm/dns"
+	"github.com/giantswarm/azure-operator/service/key"
+	"github.com/giantswarm/azuretpr"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/framework"
+)
+
+// ApplyDeleteChange deletes the resource group via the Azure API.
+func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, change interface{}) error {
+	o, err := toCustomObject(obj)
+	if err != nil {
+		return microerror.Maskf(err, "deleting host cluster DNS records")
+	}
+	c, err := toDNSRecords(change)
+	if err != nil {
+		return microerror.Maskf(err, "deleting host cluster DNS records")
+	}
+
+	return r.applyDeleteChange(ctx, o, c)
+}
```

Comment:
@xh3b4sd I like having type assertions in separation and then call private method with non-interface{} types. That makes it easier to filter this boilerplate. I'd like to do the same for New*Patch methods. I'd like this to become our "best practice". Let's discuss here.

Reply:
I think this is backwards. Having the common ground being abstracted and not polluted with custom code is a good thing. I would like to aim for some more abstraction and even code generation in the long run and I think your suggestion would not ease that but create more rules. So far we agreed on keeping the common ground as broad as possible and I would like to keep that. You can then do all custom magic in the private methods. The boilerplate you need to filter is always the first block. I do not see an issue with this personally. When you are big on this go ahead and lets see how it works out in this project but I would like to keep the majority of operators as they are for now until we have a stronger agreement.

Reply:
> I would like to aim for some more abstraction and even code generation in the long run and I think your suggestion would not ease that but create more rules. 

I could imagine generated code could look like:

```
func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
	o, err := toCustomObject(obj)
	if err != nil {
		return nil, microerror.Maskf(err, "GetDesiredState")
	}

	return r.getDesiredState(ctx, o)
}
```

You'd just have to implement `getDesiredState`.

Reply:
But this is just one part of the puzzle and the missing rest does not easily work that way because how should you guess the types being required? And doing it half that way and half another way is not going to work out.

Reply:
Types can be passed as arguments.

What would you like to generate then?

Reply:
Things like this for instance. 

```
func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
	create, err := r.newCreateChange(ctx, obj, currentState, desiredState)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	update, err := r.newUpdateChange(ctx, obj, currentState, desiredState)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	patch := framework.NewPatch()
	patch.SetCreateChange(create)
	patch.SetUpdateChange(update)

	return patch, nil
}
```

This is super generic without anything special. Guessing the private "change" methods is easy because this is a strict pattern and the outcome is predictable because it is always the same based on the private "change" methods that are implemented. Generating things like this is not going to work when we put the custom type assertion in their like you suggest, because where does the knowledge of the types come from and how does the naming of the methods than work? I think this makes it just more complicated.

Reply:
We were discussing already this can't be generalized as sometimes update means creation & deletion. This is when things are immutable.

I thought of code generation as a compensation of lack of generic types in Go. So we had very different image of that.

Reply:
I extracted `newUpdatePatch` and `newDeletePatch` which are as simple as possible, but have proper types.

I don't want to spend more time on that. I can get rid of all the private methods if that's the only thing missing to get an approval.